File: lib/core/utils/schedule_builder.dart
import '../models/installment.dart';

/// Frontend-only instalment schedule generation, isolated for the same
/// reason as LoanCalculator: a future backend owns this for real.
class ScheduleBuilder {
  ScheduleBuilder._();

  /// Builds a fully-pending instalment schedule for a new loan, spaced out
  /// from [disbursementDate] according to [frequency].
  static List<Installment> build({
    required String loanId,
    required DateTime disbursementDate,
    required int totalInstallments,
    required double installmentAmount,
    required String frequency, // daily, weekly, monthly
  }) {
    return List.generate(totalInstallments, (i) {
      final number = i + 1;
      return Installment(
        id: '$loanId-I${number.toString().padLeft(3, '0')}',
        number: number,
        dueDate: _dueDate(disbursementDate, frequency, number),
        amount: installmentAmount,
        status: InstallmentStatus.pending,
      );
    });
  }

  static DateTime _dueDate(DateTime start, String frequency, int number) {
    switch (frequency.toLowerCase()) {
      case 'daily':
        return start.add(Duration(days: number));
      case 'weekly':
        return start.add(Duration(days: 7 * number));
      case 'monthly':
      default:
        return DateTime(start.year, start.month + number, start.day);
    }
  }
}

File: backend/src/services/scheduleBuilder.js
/**
 * Installment schedule generation. Ported 1:1 from
 * lib/core/utils/schedule_builder.dart.
 */

function pad3(n) {
  return String(n).padStart(3, '0');
}

function dueDate(start, frequency, number) {
  const d = new Date(start);
  switch ((frequency || '').toLowerCase()) {
    case 'daily':
      d.setDate(d.getDate() + number);
      return d;
    case 'weekly':
      d.setDate(d.getDate() + 7 * number);
      return d;
    case 'monthly':
    default:
      return new Date(d.getFullYear(), d.getMonth() + number, d.getDate());
  }
}

/**
 * Builds a fully-pending installment schedule for a new loan.
 * @returns {Array<{id: string, loanId: string, number: number, dueDate: Date, amount: number, status: 'pending'}>}
 */
function build({ loanId, disbursementDate, totalInstallments, installmentAmount, frequency }) {
  const installments = [];
  for (let i = 0; i < totalInstallments; i++) {
    const number = i + 1;
    installments.push({
      id: `${loanId}-I${pad3(number)}`,
      loanId,
      number,
      dueDate: dueDate(disbursementDate, frequency, number),
      amount: installmentAmount,
      status: 'pending',
    });
  }
  return installments;
}

module.exports = { build };
